### ISSUE TITLE
[CreateSiFiveMetadata][firtool] Memory metadata isn't emitted without "--split-verilog" option

### ISSUE THREAD

**Issue by tymcauley** [llvm/circt#4111]
CIRCT version: `72f4c3f7`

### Issue

I'm trying to test out generating Rocket Chip RTL with CIRCT (testing out this PR https://github.com/ucb-bar/chipyard/pull/1239), and I noticed some strange behavior while trying to toggle between enabling the `firtool` option `--split-verilog` and disabling it: when that option is present, the `--repl-seq-mem-file=<PATH-TO-FILE>` option functions as expected, and we get an output file with memory metadata. However, when `--split-verilog` is removed, the memory metadata file isn't generated.

I'm invoking `firtool` in these two ways:

```
firtool \
    --format=fir \
    -O=release \
    --emit-metadata \
    <many more options ...> \
    --repl-seq-mem \
    --repl-seq-mem-circuit=TestHarness \
    --repl-seq-mem-file=./RocketConfig.mems.conf \
    --annotation-file=../RocketConfig.sfc.anno.json \
    --split-verilog \
    -o ./ \
    path/to/file.fir
```
vs
```
﻿firtool \
    --format=fir \
    -O=release \
    --emit-metadata \
    <many more options ...> \
    --repl-seq-mem \
    --repl-seq-mem-circuit=TestHarness \
    --repl-seq-mem-file=./RocketConfig.mems.conf \
    --annotation-file=../RocketConfig.sfc.anno.json \
    -o ./output.sv \
    path/to/file.fir
```

### Expected Behavior

The file specified by `--repl-seq-mem-file` should be generated, regardless of the presence of the `--split-verilog` flag.

### Reproducing

I'm attaching an archive which contains the input annotation and FIR files, as well as a script to run the test. To run the test, invoke:

```
./run-test.sh
```

Make sure `firtool` is on your PATH when you run this. The test will invoke `firtool` twice, once with `--split-verilog`, and once without. The output with `--split-verilog` will end up in `./output-dir-split`, and the output without `--split-verilog` will end up in `./output-dir-no-split`.

You should notice that the file specified in `--repl-seq-mem-file` gets generated in `./output-dir-split`:

```
$ cat output-dir-split/RocketConfig.mems.conf
name cc_dir_ext depth 1024 width 136 ports mrw mask_gran 17
name cc_banks_0_ext depth 16384 width 64 ports rw
name data_arrays_0_ext depth 512 width 256 ports mrw mask_gran 8
name tag_array_ext depth 64 width 88 ports mrw mask_gran 22
name tag_array_0_ext depth 64 width 84 ports mrw mask_gran 21
name data_arrays_0_0_ext depth 512 width 128 ports mrw mask_gran 32
name l2_tlb_ram_0_ext depth 1024 width 44 ports rw
name mem_ext depth 512 width 64 ports mrw mask_gran 8
```

However, that file isn't present in `output-dir-no-split`:

```
$ ls output-dir-no-split
RocketConfig.sv
```

Before re-running the test, make sure to remove `﻿output-dir-no-split` and `output-dir-split`.

[circt-test.tar.gz](https://github.com/llvm/circt/files/9790916/circt-test.tar.gz)

**Comment by seldridge** [comment]
I think the metadata _is_ generated, but it is inside `output.sv`. It likely looks like:

```
module Foo();
endmodule

// ---- FILE metadata.json

[ ... ]
```

Currently, any flow that is not `-split-verilog -o=outputDir` is not well supported for generating separate metadata files. There are not-great reasons for this (e.g., it makes testing easier) and we should probably reconsider it. 

As a temporary solution, you can likely throw the output Verilog into a script to split it up. I am already doing this for the single-file output case to support the Trace API in `chisel-circt` where the output annotation file (which only exists in support of the Trace API) needs to be reprocessed: https://github.com/sifive/chisel-circt/pull/25/files#diff-d0d2e73da41ac3f6492a73cf9c486d8c51ee12ae61abb78f293e626d3fb93cd0R41

A better solution would be to have `ExportVerilog` respect exactly what a user says if they want a file to be written out as opposed to changing this based on the existence of `-split-verilog` or its absence.

**Comment by tymcauley** [comment]
Yup, you're totally right, all of the metadata files are appended to the bottom of the single output file. This behavior feels a bit unexpected to me, given that the `--split-verilog` help text says: `- Emit Verilog (one file per module; specify directory with -o=<dir>)`. I was expecting that option to _only_ affect Verilog emission, not metadata emission (similar to the SFC `--emit-modules` flag).

It seems like we might not be giving `firtool` enough information, since `-o` is a bit overloaded. In the SFC, there are separate `--target-dir` and `--output-file` options, which (in my opinion) makes the intended behavior a bit easier to specify/predict. Should we consider updating `firtool` file emission arguments? Not sure if the SFC behavior is the best solution, but at least between current `firtool` and SFC behavior, the SFC route feels a bit more predictable/explicit.

**Comment by seldridge** [comment]
The way I'd think about the architecture is that no pass ever writes files. Passes that want to produce files generate verbatim text inside an `sv::VerbatimOp` and add an `output_file` attribute to it. When running with `-split-verilog` two things happen: (1) all hardware modules are given an `output_file` attribute and (2) the `output_file` attribute is used to write files during emission. When running without `-split-verilog`: (1) no `output_file` attributes are added automatically (though they may exist due to earlier passes) and (2) the `output_file` attribute is not used to generate files and instead produces the `// ---- 8> ----- FILE` separators.

It is likely cleaner to change this so that `output_file` is always respected during emission, and `-split-verilog` is instead a pre-pass during `ExportVerilog` to add `output_file` attributes to hardware modules.

Having the ability to force single-file emission with these `// ----- 8> ----- FILE` separators is useful (e.g., it facilitates testing and lets you actually serialize these to a single file or pipe for later processing while not losing file information).

From what I gather, these were the same thoughts that @fabianschuiki had when he originally wrote this. 😉  See: https://github.com/llvm/circt/pull/973

@fabianschuiki: WDYT?

**Comment by mikeurbach** [comment]
> This would slow down emission though, as we'd have to assemble potentially 100s of MB of output text in memory and uniquify them into the MLIR context as part of the op, only to flush them out to disk later on.

Drive by, haven't thought about this deeply, just the statement above: I wonder if we could use the newish MLIR "resources" for these `circt.file`s. This sounds like the kind of thing resources are intended to avoid (i.e. not forcing you to construct large StringAttrs). Also, 100s of MB of output text is probably an underestimate in some cases 😬

**Comment by fabianschuiki** [comment]
The resources mechanism sounds pretty interesting! That's probably worth looking into. I also remember @darthscsi suggesting at some point that we could have something like a `circt.file_on_disk` op that would represent an output file, but instead of having the contents live in the MLIR context, it would be in some temporary file on disk. Then the emission would copy/move/concat the files into the right final locations. That might also be interesting. But that would also subvert @seldridge's point about us _not_ wanting to have passes write to disk on their own. And yeah you're totally right, 100s of MB is probably terribly naive :stuck_out_tongue:
